emu/src/single_data_transfer.rs

```diff
@@ -0,0 +1,165 @@
+use crate::{arm7tdmi::Arm7tdmi, bitwise::Bits, io_device::IoDevice, opcode::ArmModeOpcode};
+
+/// Possible opeartion on transfer data.
+#[derive(PartialEq)]
+enum SingleDataTransfer {
+    /// Load from memory into a register.
+    Ldr,
+
+    /// Store from a register into memory.
+    Str,
+    Pld,
+}
+
+impl From<ArmModeOpcode> for SingleDataTransfer {
+    fn from(op_code: ArmModeOpcode) -> Self {
+        let must_for_pld = op_code.are_bits_on(28..=31);
+        if op_code.get_bit(20) {
+            if must_for_pld {
+                Self::Pld
+            } else {
+                Self::Ldr
+            }
+        } else {
+            Self::Str
+        }
+    }
+}
+
+/// There two different kind of write or read for memory.
+#[derive(Default)]
+enum ReadWriteKind {
+    /// Word is a u32 value for ARM mode and u16 for Thumb mode.
+    #[default]
+    Word,
+
+    /// Byte is a u8 value.
+    Byte,
+}
+
+impl From<bool> for ReadWriteKind {
+    fn from(value: bool) -> Self {
+        if value {
+            Self::Byte
+        } else {
+            Self::Word
+        }
+    }
+}
+
+impl From<&ArmModeOpcode> for ReadWriteKind {
+    fn from(op_code: &ArmModeOpcode) -> Self {
+        op_code.get_bit(22).into()
+    }
+}
+
+impl Arm7tdmi {
+    pub(crate) fn single_data_transfer(&mut self, op_code: ArmModeOpcode) -> bool {
+        let immediate = op_code.get_bit(25);
+        let up_down = op_code.get_bit(23);
+        let byte_or_word: ReadWriteKind = (&op_code).into();
+
+        // bits [19-16] - Base register
+        let rn = op_code.get_bits(16..=19);
+
+        // 0xF is register of PC
+        let address = if rn == 0xF {
+            let pc: u32 = self.registers.program_counter().try_into().unwrap();
+            pc + 8_u32
+        } else {
+            self.registers.register_at(rn.try_into().unwrap())
+        };
+
+        // bits [15-12] - Source/Destination Register
+        let rd = op_code.get_bits(12..=15);
+
+        let offset = if immediate {
+            todo!()
+        } else {
+            op_code.get_bits(0..=11)
+        };
+
+        let load_store: SingleDataTransfer = op_code.into();
+
+        let address = if up_down {
+            address.wrapping_sub(offset)
+        } else {
+            address.wrapping_add(offset)
+        };
+
+        match load_store {
+            SingleDataTransfer::Ldr => match byte_or_word {
+                ReadWriteKind::Byte => self
+                    .registers
+                    .set_register_at(rd.try_into().unwrap(), self.memory.read_at(address) as u32), // FIXME: writing a byte into u32 overwrite all register or just 0-7 bits?
+                ReadWriteKind::Word => todo!(),
+            },
+            SingleDataTransfer::Str => {
+                match byte_or_word {
+                    ReadWriteKind::Byte => self.memory.write_at(address, rd as u8), // FIXME: Is it right to truncate rd to u8?
```

Comment:
Documentation says

> A byte store (STRB) repeats the bottom 8 bits of the source register four times across data bus outputs 31 through 0. The external memory system should activate the appropriate byte subsystem to store the data.

But I don’t know what this means :)

Reply:
I think it is correct that it truncates to the first 8 bits but then I don’t know what it means that it repeats the 8 bits 4 times in the data bus.

Reply:
I also read this part, and I think that the `external memory`, in the future, should activate a byte to copy this 8 bits four times to the data bus outputs. At this moment we could ignore this part, and add a comment/issue. Right?

Reply:
But what does it mean to copy the 8 bits four times in the data bus? Do you maybe have some documentation where it is explained how the external memory reads from the data bus?

Reply:
I don't know what is this data bus, but if this is an option that we can activate in future, I will investigate on this